forte/mrdsrg-spin-integrated/dsrg_mrpt2_grad/dsrg_mrpt2_deriv_solve_zvector.cc

```diff
@@ -0,0 +1,1504 @@
+                temp_1["ijab"] += Kappa["ijab"] * Delta2["ijab"];
+                temp_1["iJaB"] += Kappa["iJaB"] * Delta2["iJaB"];
+                val["m"] += 4.0 * s_ * Tau2["mjab"] * temp["mjab"];
+                val["m"] += 8.0 * s_ * Tau2["mJaB"] * temp["mJaB"];
+                val["m"] -= 4.0 * s_ * temp["mlcd"] * temp_1["mlcd"];
+                val["m"] -= 8.0 * s_ * temp["mLcD"] * temp_1["mLcD"];
+                val["e"] -= 4.0 * s_ * Tau2["ijeb"] * temp["ijeb"];
+                val["e"] -= 8.0 * s_ * Tau2["iJeB"] * temp["iJeB"];
+                val["e"] += 4.0 * s_ * temp["kled"] * temp_1["kled"];
+                val["e"] += 8.0 * s_ * temp["kLeD"] * temp_1["kLeD"];
+                val["u"] += 4.0 * s_ * Tau2["ujab"] * temp["ujab"];
+                val["u"] += 8.0 * s_ * Tau2["uJaB"] * temp["uJaB"];
+                val["u"] -= 4.0 * s_ * temp["ulcd"] * temp_1["ulcd"];
+                val["u"] -= 8.0 * s_ * temp["uLcD"] * temp_1["uLcD"];
+                val["u"] -= 4.0 * s_ * Tau2["ijub"] * temp["ijub"];
+                val["u"] -= 8.0 * s_ * Tau2["iJuB"] * temp["iJuB"];
+                val["u"] += 4.0 * s_ * temp["klud"] * temp_1["klud"];
+                val["u"] += 8.0 * s_ * temp["kLuD"] * temp_1["kLuD"];
+            }
+        }
+    }
+
+    std::map<string, string> capital_blocks = {{"cc", "CC"}, {"vv", "VV"}, {"aa", "AA"}};
+    auto blocklabels = {"cc", "vv", "aa"};
+    std::map<char, int> orbital_size = {{'c', ncore}, {'a', na}, {'v', nvirt}};
+    for (const std::string& block : blocklabels) {
+        char label = block[1];
+        std::string slabel(1, label);
+        auto block_data =val.block(slabel).data();
+        if (label != 'a') {
+            (Z.block(block)).iterate([&](const std::vector<size_t>& i, double& value) {
+                if (i[0] == i[1]) {
+                    value = block_data[i[0]];
+                } else {
+                    auto dmt = Delta1.block(block).data()[i[1] * orbital_size[label] + i[0]];
+                    if (std::fabs(dmt) > 1e-12) {
+                        value /= dmt;
+                    }
+                }
+            });
+        } else {
+            (Z.block(block)).iterate([&](const std::vector<size_t>& i, double& value) {
+                if (i[0] == i[1]) {
+                    value = block_data[i[0]];
+                }
+            });
+        }
+    }
+    // Z[Beta] = Z[Alpha]
+    for (const std::string& block : blocklabels) {
+        Z.block(capital_blocks[block])("pq") = Z.block(block)("pq");
+    }
+    outfile->Printf("Done");
+}
+
+double f_norm(std::vector<double> const& vec) {
+    double accum = 0.0;
+    for (int i = 0; i < vec.size(); ++i) {
+        accum += vec[i] * vec[i];
+    }
+    return std::sqrt(accum);
+}
+
+double diff_f_norm(std::vector<double> const& vec1, std::vector<double> const& vec2) {
+    double accum = 0.0;
+    if (vec1.size() != vec2.size()) {
+        exit(1);
```

Comment:
Here we need to throw and exception instead of exiting and explain why we throw the exception so that a programmer /user can fix the issue.